util/hot_backup.cc

```diff
@@ -0,0 +1,654 @@
+// Version 2.0 (the "License"); you may not use this file
+// except in compliance with the License.  You may obtain
+// a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+// -------------------------------------------------------------------
+
+#include <unistd.h>
+
+#include "leveldb/atomics.h"
+#include "leveldb/env.h"
+
+#include "db/dbformat.h"
+#include "db/db_impl.h"
+#include "db/filename.h"
+#include "util/db_list.h"
+#include "util/hot_backup.h"
+#include "util/hot_threads.h"
+
+namespace leveldb {
+
+
+// this local static is used during normal operations.  unit tests
+//  create independent object.
+static HotBackup LocalHotBackupObj;
+HotBackup * gHotBackup(&LocalHotBackupObj);
+void HotBackup::ResetHotBackupObject() {gHotBackup=&LocalHotBackupObj;};
+
+
+/**
+ * Called by throttle.cc's thread once a minute.  Used to
+ *  test for trigger condition
+ */
+void
+CheckHotBackupTrigger()
+{
+    // add_and_fetch for memory fences
+    if (0 == gHotBackup->GetJobsPending())
+    {
+        if (gHotBackup->IsTriggerSet())
+        {
+            // two instance counts held for "trigger management"
+            //  (protects against hot backups happening as fast as
+            //   ScanDB() makes calls ... causing trigger to reset too
+            //   early)
+            gHotBackup->HotBackupScheduled();
+            gHotBackup->HotBackupScheduled();
+
+            // Log() routes to syslog as an error
+            Log(NULL, "leveldb HotBackup triggered.");
+            gPerfCounters->Inc(ePerfBackupStarted);
+
+            DBList()->ScanDBs(true,  &DBImpl::HotBackup);
+            DBList()->ScanDBs(false, &DBImpl::HotBackup);
+
+            // reduce to one instance count now that events posted
+            //  (trigger reset now enabled)
+            gHotBackup->HotBackupFinished();
+        }   // if
+    }   // if
+
+    return;
+
+}   // CheckHotBackupTrigger
+
+
+/**
+ * Look for external trigger
+ */
+bool
+HotBackup::IsTriggerSet()
+{
+    bool ret_flag;
+    int ret_val;
+
+    //
+    // This is code polls for existance of /etc/riak/hot_backup
+    //
+    ret_val=access(GetTriggerPath(), F_OK);
+    ret_flag=(-1!=ret_val);
+
+    return(ret_flag);
+
+}   // HotBackup::IsTriggerSet
+
+
+/**
+ * Let external trigger know that hot backup is complete.  Put
+ *  message to syslog if unable to delete trigger file
+ */
+void
+HotBackup::ResetTrigger()
+{
+    bool ret_flag;
+    int ret_val;
+
+    ret_val=unlink(GetTriggerPath());
+    if (0==ret_val)
+    {
+        // release last counter, enabling future backups
+        HotBackupFinished();
```

Comment:
HotBackupFinished() calls this method when m_JobsPending goes to 1. I'm unclear why we need that circularity.